lib/sqlalchemy/dialects/postgresql/base.py

```diff
@@ -2307,6 +2316,29 @@ def visit_create_index(self, create, **kw):
 
         return text
 
+    def visit_unique_constraint(self, constraint, **kw):
+        if len(constraint) == 0:
+            return ""
+        nulls_not_distinct_option = ""
+        nulls_not_distinct = constraint.dialect_options["postgresql"][
+            "nulls_not_distinct"
+        ]
+        if nulls_not_distinct is True:
+            nulls_not_distinct_option = "NULLS NOT DISTINCT "
+        elif nulls_not_distinct is False:
+            nulls_not_distinct_option = "NULLS DISTINCT "
+        text = ""
+        if constraint.name is not None:
+            formatted_name = self.preparer.format_constraint(constraint)
+            if formatted_name is not None:
+                text += "CONSTRAINT %s " % formatted_name
+        text += "UNIQUE %s(%s)" % (
+            nulls_not_distinct_option,
```

Comment:
**mike bayer** (zzzeek) wrote:

OK.    So....this method is copying a *lot* from compiler and NULLS NOT DISTINCT is in the SQL standard...so while I dont yet want to generalize this completely, can we do this instead:


    diff --git a/lib/sqlalchemy/sql/compiler.py b/lib/sqlalchemy/sql/compiler.py
    index d4809ef934..24065d2049 100644
    --- a/lib/sqlalchemy/sql/compiler.py
    +++ b/lib/sqlalchemy/sql/compiler.py
    @@ -6882,7 +6882,8 @@ class DDLCompiler(Compiled):
                formatted_name = self.preparer.format_constraint(constraint)
                if formatted_name is not None:
                    text += "CONSTRAINT %s " % formatted_name
    -        text += "UNIQUE (%s)" % (
    +        text += "UNIQUE %s(%s)" % (
    +            self.define_unique_constraint_distinct(constraint, **kw),
                ", ".join(self.preparer.quote(c.name) for c in constraint)
            )
            text += self.define_constraint_deferrability(constraint)


that is, put it in compiler.py as a new hook, in the same way as define_constraint_deferrability.  since this will someday be a standard hook for other backends also that want to support this.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4632

Reply:
I thought it's normal because there is similar coping in mssql dialect) 
But I agree with you that this is redundant logic. I'll fix it

Reply:
Done